### ISSUE TITLE
clang-tidy-14 findings of Meson based project not shown in "files changed" tab of pull request

### ISSUE THREAD

**Issue by lurtz** [cpp-linter/cpp-linter#30]
When adding cpp-linter to my repo I noticed that in the console output it finds all issues I added for testing, but these clang-tidy findings are not shown in the Github UI in the "files changed" tab at all. After I enabled thread-comments the findings are shown in the conversation of the pull request.

https://github.com/lurtz/sleep-proxy/pull/6

I guess the reason is using clang-tidy-14. I first used clang-tidy-12 but that had issues with finding system headers and thus provided many false positives.

**Comment by 2bndy5** [comment]
I think the solution is to adjust

https://github.com/cpp-linter/cpp-linter/blob/35e64f8fc02fe982294e940cbba9e79fd5da328c/cpp_linter/clang_tidy.py#L43-L45

to resolve the intermitten `..` into an absolute path, like so:
```py
        self.filename = (
            Path(self.filename).resolve().as_posix().replace(Path.cwd().as_posix(), "")
        )
```

But I don't have time to test this right now. I didn't know about the build system's odd path formation, that's also helpful to know when writing the unit test for this case.

**Comment by 2bndy5** [comment]
@lurtz I'm curious if Meson uses absolute or relative paths in compile_commands,json?

I've tested this using CMake on out-of-source builds, I but didn't encounter this problem. I'm still not sure if it specific to clang-tidy v14 (last tested locally using v12).

**Comment by 2bndy5** [comment]
The fact that the paths start with a `/` is a serious problem because technically that is an absolute path.

- `build/../src/blah` is a relative path with relative redirects (`..`)
- `/build/../src/blah` is an absolute path with relative redirects (`..`)

Is there a way to customize Meson to avoid this? I've never heard or used that build system generator.

**Comment by 2bndy5** [comment]
@lurtz  I found this to be informative https://github.com/mesonbuild/meson/issues/10227

**Comment by lurtz** [comment]
That could explain why it clang-tidy versions prior 14 did not work for me.

This is a compile_commands.json file generated on my notebook:
[compile_commands.json.txt](https://github.com/cpp-linter/cpp-linter/files/10776909/compile_commands.json.txt)
As you can see command, file and output are all relative to directory for each translation unit.

I will create a toy C++ project with CMake and Meson to see the differences in the generated compile_commands.json.

**Comment by lurtz** [comment]
> The fact that the paths start with a `/` is a serious problem because technically that is an absolute path.
> 
>     * `build/../src/blah` is a relative path with relative redirects (`..`)
> 
>     * `/build/../src/blah` is an absolute path with relative redirects (`..`)
> 
> 
> The fix I proposed above will not recognize `/build/../src/` as a relative path as you can see from the following interactive python session:
> 
> ```python
> >>> from pathlib import PurePath
> >>> PurePath("/build/../src").is_absolute()
> True
> >>> PurePath("build/../src").is_absolute()
> False
> ```

Maybe it is enough to remove the "dir/.." pattern. I got it working with the following Python code:

    In [1]: from pathlib import Path
    In [2]: Path('/foo/bar/../src').resolve()
    Out[2]: PosixPath('/foo/src')
    In [3]: Path('foo/bar/../src').resolve()
    Out[3]: PosixPath('/home/lutz/workspace/sleep-proxy/.github/workflows/foo/src')


I checked the console output of the last failing and then successful build (where I tweak the compile_commands.json so that everthing starts from the source directory).

[No annotations:](https://github.com/lurtz/sleep-proxy/actions/runs/4211802730/jobs/7310386899)

    INFO:CPP Linter:Running "clang-tidy-14 --export-fixes=clang_tidy_output.yml -p /home/runner/work/sleep-proxy/sleep-proxy/build --extra-arg= src/emulateHost.cpp"
    DEBUG:CPP Linter:Output from clang-tidy:
    ../src/emulateHost.cpp:28:4: warning: unused variable 'x' [clang-diagnostic-unused-variable]
              x
              ^
    /home/runner/work/sleep-proxy/sleep-proxy/build/../src/emulateHost.cpp:32:7: warning: redundant boolean literal in if statement condition [readability-simplify-boolean-expr]
      if (false) {
      ~~~~^~~~~~~~
    /home/runner/work/sleep-proxy/sleep-proxy/build/../src/emulateHost.cpp:33:18: warning: variable 'sptr' is not initialized [cppcoreguidelines-init-variables]
              std::string * sptr;
                            ^
    note: this fix will not be applied because it overlaps with another fix
    /home/runner/work/sleep-proxy/sleep-proxy/build/../src/emulateHost.cpp:40:15: warning: do not use pointer arithmetic [cppcoreguidelines-pro-bounds-pointer-arithmetic]
        setup_log(argv[0], 0, LOG_DAEMON);
                  ^

[With annotations:](https://github.com/lurtz/sleep-proxy/actions/runs/4213473474/jobs/7313201206)

    INFO:CPP Linter:Running "clang-tidy-14 --export-fixes=clang_tidy_output.yml -p /home/runner/work/sleep-proxy/sleep-proxy/build --extra-arg= src/emulateHost.cpp"
    DEBUG:CPP Linter:Output from clang-tidy:
    src/emulateHost.cpp:28:4: warning: unused variable 'x' [clang-diagnostic-unused-variable]
              x
              ^
    /home/runner/work/sleep-proxy/sleep-proxy/src/emulateHost.cpp:32:7: warning: redundant boolean literal in if statement condition [readability-simplify-boolean-expr]
      if (false) {
      ~~~~^~~~~~~~
    /home/runner/work/sleep-proxy/sleep-proxy/src/emulateHost.cpp:33:18: warning: variable 'sptr' is not initialized [cppcoreguidelines-init-variables]
              std::string * sptr;
                            ^
    note: this fix will not be applied because it overlaps with another fix
    /home/runner/work/sleep-proxy/sleep-proxy/src/emulateHost.cpp:40:15: warning: do not use pointer arithmetic [cppcoreguidelines-pro-bounds-pointer-arithmetic]
        setup_log(argv[0], 0, LOG_DAEMON);
                  ^

You can see in both cases absolute paths are used, but in the non working case a detour via "build/.." is taken.

**Comment by lurtz** [comment]
> I will create a toy C++ project with CMake and Meson to see the differences in the generated compile_commands.json.

compile_commands.json with CMake and Ninja:

    [
    {
      "directory": "/home/lutz/workspace/tiny-cpp-project/build",
      "command": "/usr/bin/c++ -o CMakeFiles/hello_world.dir/src/main.cpp.o -c /home/lutz/workspace/tiny-cpp-project/src/main.cpp",
      "file": "/home/lutz/workspace/tiny-cpp-project/src/main.cpp"
    }

compile_commands.json with Meson and Ninja:

    [
      {
        "directory": "/home/lutz/workspace/tiny-cpp-project/build_meson",
        "command": "c++ -Ihello_world.p -I. -I.. -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -O0 -g -MD -MQ hello_world.p/src_main.cpp.o -MF hello_world.p/src_main.cpp.o.d -o hello_world.p/src_main.cpp.o -c ../src/main.cpp",
        "file": "../src/main.cpp",
        "output": "hello_world.p/src_main.cpp.o"
      }
    ]

So CMake always uses absolute paths. I do not know if Ninja is to blame, because I used that with CMake as well. Maybe CMake does not use Ninja to generate the compile_commands.json. This evening I will check how CMake and Meson are storing the paths internally.

**Comment by 2bndy5** [comment]
> I do not know if Ninja is to blame, because I used that with CMake as well. Maybe CMake does not use Ninja to generate the compile_commands.json.

This is a good observation. The discrepancy then is likely in the build generator's input to the tool(s) that create the database. 

### Looking at the CMake source
It isn't immediately clear how this is done with ninja; [they set certain target properties](https://gitlab.kitware.com/cmake/cmake/-/blob/master/Source/cmNinjaTargetGenerator.cxx#L1755) when the target property `EXPORT_COMPILE_COMMANDS` is asserted.

### Looking at the Meson source
It is a little clearer how this is done, but I don't know if [the args passed to `ninja -t compdb`](https://github.com/mesonbuild/meson/blob/019fda91d0cb4308803c32e8d9dd311b85caf702/mesonbuild/backend/ninjabackend.py#L667) are the point of error.

**Comment by lurtz** [comment]
You are right Meson is using relative paths. Look at the build.ninja file.
CMake:

    lutz@brutus:~/workspace/tiny-cpp-project/build$ grep src build.ninja 
    build CMakeFiles/hello_world.dir/src/main.cpp.o: CXX_COMPILER__hello_world_ /home/lutz/workspace/tiny-cpp-project/src/main.cpp || cmake_object_order_depends_target_hello_world

Meson:

    lutz@brutus:~/workspace/tiny-cpp-project/build_meson$ grep src build.ninja 
    build hello_world.p/src_main.cpp.o: cpp_COMPILER ../src/main.cpp

However I cannot judge if using relative paths is wrong. I do not know the format specification of compile_commands.json.

**Comment by 2bndy5** [comment]
> Tools which always expect absolute paths are too strict.

clang-tools are famous for their strictness.

**Comment by 2bndy5** [comment]
And now for the bad news...

The way I see it, it is not within this action's scope to post-process a compilation database. Thus I don't see a way that we can address this issue from our side -- not unless Meson is persuaded to use absolute paths when generating the ninja build file(s).

On a side note, it may be within the specifications to use relative paths, but absolute paths are preferred in many best practiced conventions. We also made the move to ensure the path specified to our `database` option was resolved to an absolute path (mostly because of troubles with historically using a docker container).

**Comment by lurtz** [comment]
You could add some documentation. E.g. meson users have to use the database option to specify the location of the file. That could be a place to tell Meson users that they need at least clang-tidy-14 and that they have to modify the compile_commands.json so that e.g. in no path repoorted by clang-tidy the pattern  "../src" appears.

I got it working with two sed commands and helpful documentation would prevent people opening issues in the future.

**Comment by 2bndy5** [comment]
The main branch (as of v1.6.5) now uses the specified database path to load the JSON database and resolve relative paths automatically. @lurtz The manual "massaging" of the database with `sed` should no longer be needed.

FYI, it turns out that ninja "prefers relative paths whenever possible".
